src/index.ts

```diff
@@ -0,0 +1,1368 @@
+                // which is the corresponding event name
+                p["event-handler"] &&
+                // The list is partly obtained from the table at
+                // http://www.w3.org/TR/DOM-Level-3-Events/#dom-events-conformance   #4.1
+                eNameToEType[p["event-handler"]!] || defaultEventType;
+
+            return eType === "Event" || dependsOn(eType, "Event")
+                ? eType
+                : defaultEventType;
+        }
+
+        // Get all the event handlers from an interface and also from its inherited / implemented interfaces
+        function getEventHandler(i: Browser.Interface) {
+            const ownEventHandler =
+                i.properties
+                    ? mapToArray(i.properties.property).filter(p => p["event-handler"]).map(p => ({
+                        name: p.name,
+                        eventName: p["event-handler"]!,
+                        eventType: getEventTypeFromHandler(p)
+                    }))
+                    : [];
+            return ownEventHandler;
+        }
+
+        const result: Record<string, EventHandler[]> = {};
+        for (const i of allInterfaces) {
+            result[i.name] = getEventHandler(i);
+        }
+        return result;
+    })();
+
+    // Map of interface.Name -> List of base interfaces with event handlers
+    const iNameToEhParents = (() => {
+        function hasHandler(i: Browser.Interface) {
+            return iNameToEhList[i.name] && iNameToEhList[i.name].length;
+        }
+        // Get all the event handlers from an interface and also from its inherited / implemented interfaces
+        function getParentsWithEventHandler(i: Browser.Interface) {
+            function getParentEventHandler(i: Browser.Interface): Browser.Interface[] {
+                return hasHandler(i) ? [i] : getParentsWithEventHandler(i);
+            }
+
+            const extendedParentWithEventHandler = allInterfacesMap[i.extends] && getParentEventHandler(allInterfacesMap[i.extends]) || [];
+
+            const implementedParentsWithEventHandler =
+                i.implements
+                    ? i.implements.reduce<Browser.Interface[]>((acc, i) => {
+                        acc.push(...getParentEventHandler(allInterfacesMap[i]));
+                        return acc;
+                    }, [])
+                    : [];
+
+            return extendedParentWithEventHandler.concat(implementedParentsWithEventHandler);
+        }
+
+        const result: Record<string, Browser.Interface[]> = {};
+        for (const i of allInterfaces) {
+            result[i.name] = getParentsWithEventHandler(i);
+        }
+        return result;
+    })();
+
+    return flavor === Flavor.ES6Iterators ? emitES6DomIterators() : emit();
+
+    // Used to decide if a member should be emitted given its static property and
+    // the intended scope level.
+    function matchScope(scope: EmitScope, x: Browser.Method) {
```

Comment:
Seems like this and several other functions don't depend on any state. Would be nice to move them out of the closure so it's clearer what *is* staeful.

Reply:
@mhegazy missed one